File: data_logger.py
#!/usr/bin/python

import sys
from bcc import BPF
from time import sleep
import pyroute2
import csv

def usage():
    print("Usage: {0} <ifdev>".format(sys.argv[0]))
    print("e.g.: {0} eth0\n".format(sys.argv[0]))
    exit(1)
if len(sys.argv) !=2:
    usage()

# Load the eBPF program
bpf = BPF(src_file="data_logger.c")

# Attach XDP program to a network interface for ingress traffic
device = sys.argv[1]
fn_ingress = bpf.load_func("handle_ingress", BPF.XDP)
bpf.attach_xdp(device, fn_ingress)

# Attach TC program to the same network interface for egress traffic
ipr = pyroute2.IPRoute()
ipdb = pyroute2.IPDB(nl=ipr)
idx = ipdb.interfaces[device].index
ipr.tc("add", "clsact", idx)
fn_egress = bpf.load_func("handle_egress", BPF.SCHED_CLS)
ipr.tc("add-filter", "bpf", idx, ":1", fd=fn_egress.fd, name=fn_egress.name,
       parent="ffff:fff3", classid=1, direct_action=True)

print(f"eBPF program attached to {device}. Press Ctrl+C to stop.")

# Open a CSV file for logging packet data
with open("packet_data.csv", "w") as csvfile:
    csv_writer = csv.writer(csvfile)
    csv_writer.writerow(["Timestamp (ns)", "IAT (ns)", "Packet Length", "Direction"])

    # Callback function to process events from the perf buffer
    def process_event(cpu, data, size):
        event = bpf["events"].event(data)
        direction = "Incoming" if event.direction == 0 else "Outgoing"
        csv_writer.writerow([event.timestamp_ns, event.iat_ns, event.packet_length, direction])
        print(f"{direction} Packet - Length: {event.packet_length} bytes, IAT: {event.iat_ns} ns")

    # Attach the perf buffer
    bpf["events"].open_perf_buffer(process_event)

    try:
        while True:
            bpf.perf_buffer_poll()
    except KeyboardInterrupt:
        # Clean up on exit
        print("Detaching program and exiting...")
        bpf.remove_xdp(device)
        ipr.tc("del", "clsact", idx)
        ipdb.release()

File: data_logger.c
#include <uapi/linux/bpf.h>
#include <linux/ip.h>
#include <linux/if_packet.h>
#include <uapi/linux/if_ether.h>
#include <uapi/linux/pkt_cls.h>

// Define a perf event map for streaming data to userspace
struct packet_event {
    u64 timestamp_ns;  // Current timestamp
    u64 iat_ns;        // Inter-arrival time
    u64 packet_length; // Packet length
    u8 direction;      // 0 = Incoming, 1 = Outgoing
};
BPF_PERF_OUTPUT(events);

// Define a map to store the last timestamp per CPU (for inter-arrival time)
BPF_PERCPU_ARRAY(last_timestamp, u64, 2); // Index 0: Incoming, 1: Outgoing

// Helper function to calculate inter-arrival time and stream data
static __always_inline void record_packet_event(void *ctx,u32 direction, u64 packet_length) {
    u64 now = bpf_ktime_get_ns(); // Current time in nanoseconds

    // Lookup last timestamp for the given direction
    u64 *last_time = last_timestamp.lookup(&direction);
    u64 iat = 0;
      if (last_time) {
        if (*last_time != 0) {
            iat = now - *last_time; // Calculate inter-arrival time if it's not the first packet
        }
    }
    last_timestamp.update(&direction, &now); // Update the last timestamp

    // Prepare event data
    struct packet_event event = {};
    event.timestamp_ns = now;
    event.iat_ns = iat;
    event.packet_length = packet_length;
    event.direction = direction;

    // Send the event to userspace via the perf buffer
    events.perf_submit(ctx, &event, sizeof(event));
}

// XDP Program for Incoming Packets
int handle_ingress(struct xdp_md *ctx) {
    // Packet length
    u64 len = (u64)(ctx->data_end - ctx->data);

    // Record event for incoming packets (direction = 0)
    record_packet_event(ctx,0, len);

    return XDP_PASS; // Let the packet pass
}

// TC Program for Outgoing Packets
int handle_egress(struct __sk_buff *skb) {
    // Packet length
    u64 len = skb->len;

    // Record event for outgoing packets (direction = 1)
    record_packet_event(skb,1, len);

    return TC_ACT_OK; // Let the packet pass
}
